### ISSUE TITLE
Can't use LibreOffice's Help, or HTML files with accompanying CSS files, in sandboxed browsers

### ISSUE THREAD

**Issue by wjt** [flatpak/xdg-desktop-portal#555]
Flatpak:ed LibreOffice ships an offline version of its extensive documentation in a runtime, `org.libreoffice.LibreOffice.Help`. My default web browser is also a Flatpak application. (It doesn't particularly matter which one because all of Chromium, Firefox and Epiphany exhibit all the problems described here, with one exception that I will call out.)

When I hit F1 in LibreOffice, as best I can tell it writes a temporary file to some location or other, then invokes `xdg-open` or equivalent on it. I see the OpenFile method call (with just an FD), then the browser being activated on `/run/user/1001/doc/dcca5ada/NewHelp1.html`.

That file is as follows, though I have pretty-printed it for legibility:

```html
<!DOCTYPE HTML>
<html lang="en-US">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="1; url='file:///var/lib/flatpak/runtime/org.libreoffice.LibreOffice.Help/x86_64/stable/22757a622f8d4a5cb563e0249b84d2ef425c96d4b16e3dbdb692d199ea1e086c/files/index.html?Target=swriter/SW_HID_EDIT_WIN&Language=en-GB&System=UNIX&Version=7.0'">
    <script type="text/javascript">
        window.location.href = "file:///var/lib/flatpak/runtime/org.libreoffice.LibreOffice.Help/x86_64/stable/22757a622f8d4a5cb563e0249b84d2ef425c96d4b16e3dbdb692d199ea1e086c/files/index.html?Target=swriter/SW_HID_EDIT_WIN&Language=en-GB&System=UNIX&Version=7.0";
    </script>
    <title>Help Page Redirection</title>
</head>

<body></body>

</html>
```

So, the browser opens at that file, and follows the redirect to `file:///var/lib/flatpak/runtime/org.libreoffice.LibreOffice.Help/x86_64/stable/22757a622f8d4a5cb563e0249b84d2ef425c96d4b16e3dbdb692d199ea1e086c/files/index.html?Target=swriter/SW_HID_EDIT_WIN&Language=en-GB&System=UNIX&Version=7.0`, which of course it cannot access because that path on the host is not visible in the sandbox.

One part of the issue here is that LibreOffice has this indirection through a temporary file. Flatpak is able to arrange for that path to be accessible to the target application through the document portal, but has no idea that the target application will try to open a completely different path. I assume that it does this because it wants to pass query parameters through to the target application (which is not supported by `OpenFile`), and [`OpenURI` explicitly does not support `file:///` URIs](https://flatpak.github.io/xdg-desktop-portal/portal-docs.html#gdbus-method-org-freedesktop-portal-OpenURI.OpenURI).

But even without that indirection, there is still an issue. Suppose I have the following two files, `index.html` and `index.css`, somewhere in my home directory:

```html
<!DOCTYPE HTML>
<html lang="en-US">
<head>
  <link rel="stylesheet" href="./index.css">
</head>
<body>
  <h1>Hello World</h1>
</body>
</html>
```

```css
h1 {
  color: red;
}
```

When I use Nautilus to open `index.html` in Epiphany or Firefox, they are passed a document portal path. But only `index.html` is mapped, not its sibling file, so the text is black, not red. (Chromium currently has access to home for reasons, so this can't be reproduced there.)

Some possible approaches:

- Sidestep this entirely by giving LibreOffice its own web view, which would be in the same sandbox and would be able to access the same help.
- Make LibreOffice run a web server on localhost, and have it launch that URL in the default browser (which could also avoid the need for a temporary file)
- Make it possible to open an HTML file and have the entire parent directory mapped through the document portal, not just the file itself

I'm not entirely sure what the conditions for the third approach would be – it's by design that opening a file does not share the entire directory tree in general, because then opening a file in `~` would give the app access to the whole home directory. So I guess it would need to be opt-in by the requesting application.

**Comment by wjt** [comment]
@ramcq You made a reference to a previous conversation about a `--with-folder` argument for `xdg-open` or perhaps `flatpak run --file-forwarding` that had been mooted. I can't remember where and/or with whom you had discussed that!

**Comment by Mikenux** [comment]
I don't think this is a case for neighboring files which are about the host's files, not those of sandboxed applications or a runtime. It's more of a file-application association.

Also, asking a user if they are ok with opening `index.html` and others in a web browser to open help is definitely not good UX. It's something that should be asked like this: "LibreOffice wants to open help. Allow it with the following web browser? [list of installed web browsers] [+ checkbox to remember choice]". However, to use "help" as a word, we must know that it is indeed help...

An alternative can be flatpak having a dedicated "web engine" to display documents, or use that of a runtime if there is one (for example, that of the GNOME runtime). This should be intended for offline viewing (so no network access), so apps can use it as is without asking, all in a proper interface.

**Comment by Mikenux** [comment]
I made a proposal for this: https://github.com/flatpak/xdg-desktop-portal/discussions/1392

@mjog: What is the Debian’s installed documentation? Which SDK doc? That’s for apps here, not system.

**Comment by Mikenux** [comment]
For that, a web browser can perhaps request to open a folder.

The other methods are:
- An help app like in my proposal.
- [Related Files](https://github.com/flatpak/xdg-desktop-portal/discussions/1290) or [File Set Request](https://github.com/flatpak/xdg-desktop-portal/discussions/1256), depending on file organization (Related Files are better for this: main file + related folder of the same name)

**Comment by Mikenux** [comment]
I meant from the app. There is also this PR for flatpak to open directories from the file manager: https://github.com/flatpak/flatpak/pull/6247. However, I don’t recommend this method.

What I recommend is that developers move to "related files". There is also @swick’s proposal to have folders as files (see https://github.com/flatpak/xdg-desktop-portal/discussions/1430), which I think should be standardized, so portals can handle it in specific ways (e.g. write access) if wanted. However, nothing is official. In my opinion, decisions should be made on which solutions are going to be used, and then make this official by making an announcement, so that developers can already migrate to the recommended solutions (even if they are not there yet).
